Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalTextFont.cpp

```diff
@@ -0,0 +1,404 @@
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+//  plDXTextFont Class Functions                                             //
+//  Cyan, Inc.                                                               //
+//                                                                           //
+//// Version History //////////////////////////////////////////////////////////
+//                                                                           //
+//  2.19.2001 mcn - Created.                                                 //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "plMetalTextFont.h"
+
+#include "HeadSpin.h"
+#include "hsWindows.h"
+#include "plMetalPipeline.h"
+#include "plPipeline/hsWinRef.h"
+
+// Following number needs to be at least: 64 chars max in plTextFont drawn at any one time
+//                                      * 4 primitives per char max (for bold text)
+//                                      * 3 verts per primitive
+
+// const uint32_t  kNumVertsInBuffer(32768);
+const uint32_t kNumVertsInBuffer(4608);
+
+uint32_t plMetalTextFont::fBufferCursor = 0;
+
+//// Constructor & Destructor /////////////////////////////////////////////////
+
+plMetalTextFont::plMetalTextFont(plPipeline *pipe, plMetalDevice *device) : plTextFont(pipe),
+                                                                            fTexture()
+{
+    fDevice = device;
+    fPipeline = (plMetalPipeline *)pipe;
+    CreateShared(&(fPipeline->fDevice));
+}
+
+plMetalTextFont::~plMetalTextFont()
+{
+    DestroyObjects();
+}
+
+//// ICreateTexture ///////////////////////////////////////////////////////////
+
+void plMetalTextFont::ICreateTexture(uint16_t *data)
+{
+    printf("Create texture\n");
+
+    MTL::TextureDescriptor *descriptor = MTL::TextureDescriptor::texture2DDescriptor(MTL::PixelFormatRGBA8Unorm, fTextureWidth, fTextureHeight, false);
+
+    fTexture->release();
+    fTexture = fDevice->fMetalDevice->newTexture(descriptor);
+    fTexture->setLabel(MTLSTR("Font texture"));
+
+    struct InDataValues
+    {
+        uint8_t a : 4;
+        uint8_t r : 4;
+        uint8_t g : 4;
+        uint8_t b : 4;
```

Comment:
What's going on here?

Reply:
I think this was addressed in the other comment - but in general Plasma's font texture system is a little wonky. FreeType should be rendering everything in 8 bit color but all the passes that happen before we get here mess things up. I don't know why this is coming in as 4 bit data (but I do have to convert it up to 8.)

My guess is back when Plasma was written they were being clever by saving space where they could. But modern graphics hardware does not do 4 bit color - and if the driver offers it, it's probably doing the same thing I am.